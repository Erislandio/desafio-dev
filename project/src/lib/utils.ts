import { clsx, type ClassValue } from "clsx";
import { twMerge } from "tailwind-merge";

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs));
}

function formatDate(dataStr: string) {
  const year = dataStr.slice(0, 4);
  const month = dataStr.slice(4, 6);
  const day = dataStr.slice(6, 8);
  return `${day}/${month}/${year}`;
}

function formatHour(hour: string) {
  const time = hour.slice(0, 2);
  const min = hour.slice(2, 4);
  const seg = hour.slice(4, 6);
  return `${time}:${min}:${seg}`;
}

function parseRowToObj(row: string) {
  const type = row.slice(0, 1);

  if (!type) return null;

  return {
    type,
    date: formatDate(row.slice(1, 9)),
    value: parseInt(row.slice(9, 19)) / 100,
    document: row.slice(19, 30),
    cardNumber: row.slice(30, 42).trim(),
    hour: formatHour(row.slice(42, 48)),
    owner: row.slice(48, 62).trim(),
    name: row.slice(62, 81).trim(),
  };
}

export const fileToTransactions = (file: string) => {
  const rows = file.split("\n");

  const transactions = rows.map(parseRowToObj).filter(Boolean);

  return transactions;
};

export function formatPrice(price: number): string {
  return price.toLocaleString("pt-BR", {
    style: "currency",
    currency: "BRL",
  });
}
