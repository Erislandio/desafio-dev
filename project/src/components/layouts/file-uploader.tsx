"use client";

import type * as React from "react";
import { useEffect, useRef, useState } from "react";

import { Button } from "@/components/ui/button";
import { cn, fileToTransactions } from "@/lib/utils";
import DragAndDropFile from "./file-drag-drop";
import FileList from "./files-list";
import TransactionTable from "./result-table";
import UploadProgress from "./upload-progress";

interface FileUploaderProps extends React.HTMLAttributes<HTMLDivElement> {
  onUpload?: (files: File[]) => Promise<void>;
  maxSize?: number;
}

export interface Transation {
  name: string;
  cardNumber: string;
  document: string;
  value: number;
  date: string;
  hour: string;
  type: string;
  owner: string;
  debits: number;
  credits: number;
  transactions: Transation[];
}

export function FileUploader({
  onUpload,
  className,
  maxSize = 1024,
  ...props
}: FileUploaderProps) {
  const [isDragging, setIsDragging] = useState(false);
  const [files, setFiles] = useState<File[]>([]);
  const [uploading, setUploading] = useState(false);
  const [progress, setProgress] = useState(0);
  const [error, setError] = useState<string | null>(null);
  const fileInputRef = useRef<HTMLInputElement>(null);
  const [transactions, setTransaction] = useState<Transation[]>([]);

  useEffect(() => {
    if (files.length === 0) return;

    const readAllFiles = async () => {
      const promises = files.map((file) => {
        return new Promise((resolve, reject) => {
          const reader = new FileReader();
          reader.onload = (e) => resolve(e?.target?.result);
          reader.onerror = (e) => reject(e);
          reader.readAsText(file);
        });
      });

      try {
        const results = (await Promise.all(promises)) as string[];
        const transactions = fileToTransactions(results[0]) as Transation[];

        setTransaction((prev) => [...prev, ...transactions]);
      } catch {
        setError(`Erro ao ler arquivos: ${files.pop()?.name}`);
      }
    };

    readAllFiles();
  }, [files]);

  const handleDragOver = (e: React.DragEvent<HTMLDivElement>) => {
    e.preventDefault();
    e.stopPropagation();
    setIsDragging(true);
  };

  const handleDragLeave = (e: React.DragEvent<HTMLDivElement>) => {
    e.preventDefault();
    e.stopPropagation();
    setIsDragging(false);
  };

  const validateFiles = (fileList: FileList | null): File[] => {
    if (!fileList) return [];

    setError(null);
    const validFiles: File[] = [];

    Array.from(fileList).forEach((file) => {
      if (file.size > maxSize) {
        setError(
          `o arquivo "${file.name}" exedeu o tamanho máximo de ${
            maxSize / (1024 * 1024)
          }MB`
        );
        return;
      }

      if (!file.name.endsWith(".txt") || !file.name.includes(".txt")) {
        return setError(`o Arquivo ${file.name} não é válido!`);
      }

      validFiles.push(file);
    });

    return validFiles;
  };

  const handleDrop = (e: React.DragEvent<HTMLDivElement>) => {
    e.preventDefault();
    e.stopPropagation();
    setIsDragging(false);

    const validFiles = validateFiles(e.dataTransfer.files);
    if (validFiles.length > 0) {
      setFiles((prevFiles) => [...prevFiles, ...validFiles]);
    }
  };

  const handleFileChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const validFiles = validateFiles(e.target.files);

    if (validFiles.length > 0) {
      setFiles((prevFiles) => [...prevFiles, ...validFiles]);
    }

    if (fileInputRef.current) {
      fileInputRef.current.value = "";
    }
  };

  const handleUpload = async () => {
    if (files.length === 0) return;

    setUploading(true);
    setProgress(0);

    const interval = setInterval(() => {
      setProgress((prev) => {
        const newProgress = prev + 10;
        if (newProgress >= 100) {
          clearInterval(interval);
          return 100;
        }
        return newProgress;
      });
    }, 300);

    if (onUpload) {
      await onUpload(files);
    }

    setProgress(100);

    setTimeout(() => {
      setFiles([]);
      setUploading(false);
      setProgress(0);
    }, 1000);
  };

  const removeFile = (index: number) => {
    setTransaction([]);
    setFiles((prevFiles) => prevFiles.filter((_, i) => i !== index));
  };

  return (
    <div className={cn("space-y-4", className)} {...props}>
      <DragAndDropFile
        fileInputRef={fileInputRef}
        handleDragLeave={handleDragLeave}
        handleDragOver={handleDragOver}
        handleDrop={handleDrop}
        isDragging={isDragging}
        uploading={uploading}
        handleFileChange={handleFileChange}
        disbabled={files.length > 0}
      />

      {error && (
        <div className="bg-destructive/10 text-destructive text-sm p-3 rounded-md">
          {error}
        </div>
      )}

      <FileList files={files} onRemove={removeFile} uploading={uploading} />
      <TransactionTable transactions={transactions} />

      <UploadProgress progress={progress} uploading={uploading} />

      <div className="flex justify-end">
        <Button
          onClick={handleUpload}
          disabled={files.length === 0 || uploading}
        >
          {uploading ? "Enviando..." : "Enviar"}
        </Button>
      </div>
    </div>
  );
}
