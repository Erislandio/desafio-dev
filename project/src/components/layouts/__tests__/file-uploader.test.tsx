import {
  act,
  fireEvent,
  render,
  screen,
  waitFor,
} from "@testing-library/react";
import { fileToTransactions } from "../../../lib/utils";
import { FileUploader } from "../file-uploader";

jest.mock("../../../lib/utils", () => ({
  cn: jest.fn((...args) => args.filter(Boolean).join(" ")),
  fileToTransactions: jest.fn(),
}));

jest.mock("../file-drag-drop", () => ({
  __esModule: true,
  default: ({
    fileInputRef,
    handleDragLeave,
    handleDragOver,
    handleDrop,
    isDragging,
    uploading,
    handleFileChange,
    disbabled,
  }) => (
    <div data-testid="drag-drop-component">
      <button
        data-testid="simulate-drop"
        onClick={(e) =>
          handleDrop(e as unknown as React.DragEvent<HTMLDivElement>)
        }
      >
        Simulate Drop
      </button>
      <button
        data-testid="simulate-drag-over"
        onClick={(e) =>
          handleDragOver(e as unknown as React.DragEvent<HTMLDivElement>)
        }
      >
        Simulate Drag Over
      </button>
      <button
        data-testid="simulate-drag-leave"
        onClick={(e) =>
          handleDragLeave(e as unknown as React.DragEvent<HTMLDivElement>)
        }
      >
        Simulate Drag Leave
      </button>
      <input
        type="file"
        data-testid="file-input"
        ref={fileInputRef}
        onChange={handleFileChange}
      />
      <div>isDragging: {isDragging.toString()}</div>
      <div>uploading: {uploading.toString()}</div>
      <div>disabled: {disbabled.toString()}</div>
    </div>
  ),
}));

jest.mock("../files-list", () => ({
  __esModule: true,
  default: ({ files, onRemove, uploading }) => (
    <div data-testid="files-list-component">
      {files.map((file, index) => (
        <div key={index} data-testid={`file-${index}`}>
          {file.name}
          <button
            data-testid={`remove-file-${index}`}
            onClick={() => onRemove(index)}
            disabled={uploading}
          >
            Remove
          </button>
        </div>
      ))}
    </div>
  ),
}));

jest.mock("../result-table", () => ({
  __esModule: true,
  default: ({ transactions }) => (
    <div data-testid="transaction-table-component">
      {transactions.map((transaction, index) => (
        <div key={index} data-testid={`transaction-${index}`}>
          {transaction.name}
        </div>
      ))}
    </div>
  ),
}));

jest.mock("../upload-progress", () => ({
  __esModule: true,
  default: ({ progress, uploading }) => (
    <div data-testid="upload-progress-component">
      <div>Progress: {progress}</div>
      <div>Uploading: {uploading.toString()}</div>
    </div>
  ),
}));

jest.mock("../../ui/button", () => ({
  Button: ({ onClick, disabled, children }) => (
    <button data-testid="upload-button" onClick={onClick} disabled={disabled}>
      {children}
    </button>
  ),
}));

describe("FileUploader Component", () => {
  const mockOnUpload = jest.fn(() => Promise.resolve());

  beforeEach(() => {
    jest.clearAllMocks();
    jest.useFakeTimers();

    global.FileReader = class {
      onload;
      onerror;
      readAsText(file) {
        setTimeout(() => {
          this.onload({ target: { result: "mock file content" } });
        }, 0);
      }
    } as unknown as typeof FileReader;

    (fileToTransactions as jest.Mock).mockReturnValue([
      {
        name: "Store A",
        cardNumber: "1234",
        document: "123.456.789-00",
        value: 100,
        date: "2023-01-01",
        hour: "12:00",
        type: "1",
        owner: "John Doe",
        debits: 0,
        credits: 100,
        transactions: [],
      },
    ]);
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  test("renders component with initial state", () => {
    act(() => render(<FileUploader />));

    expect(screen.getByTestId("drag-drop-component")).toBeInTheDocument();
    expect(screen.getByTestId("upload-button")).toBeInTheDocument();
    expect(screen.getByTestId("upload-button")).toBeDisabled();
    expect(screen.getByText("Enviar")).toBeInTheDocument();
  });

  test("handles drag over event", async () => {
    render(<FileUploader />);

    act(() => {
      fireEvent.click(screen.getByTestId("simulate-drag-over"));
    });

    await waitFor(() => {
      expect(screen.getByText("isDragging: true")).toBeInTheDocument();
    });
  });

  test("handles drag leave event", async () => {
    render(<FileUploader />);

    act(() => {
      fireEvent.click(screen.getByTestId("simulate-drag-over"));
    });

    await waitFor(() => {
      expect(screen.getByText("isDragging: true")).toBeInTheDocument();
    });

    act(() => {
      fireEvent.click(screen.getByTestId("simulate-drag-leave"));
    });

    await waitFor(() => {
      expect(screen.getByText("isDragging: false")).toBeInTheDocument();
    });
  });

  test("handles file input change with valid file", async () => {
    render(<FileUploader />);

    const file = new File(["file content"], "test.txt", { type: "text/plain" });

    const input = screen.getByTestId("file-input");
    act(() => {
      fireEvent.change(input, { target: { files: [file] } });
    });

    await waitFor(() => {
      expect(screen.getByTestId("file-0")).toBeInTheDocument();
      expect(screen.getByTestId("file-0")).toHaveTextContent("test.txt");
    });
  });

  test("validates file extension", async () => {
    render(<FileUploader />);

    const file = new File(["file content"], "test.pdf", {
      type: "application/pdf",
    });

    const input = screen.getByTestId("file-input");

    act(() => {
      fireEvent.change(input, { target: { files: [file] } });
    });

    await waitFor(() => {
      expect(screen.getByText(/não é válido/)).toBeInTheDocument();
    });
  });

  test("removes file when remove button is clicked", async () => {
    render(<FileUploader />);

    const file = new File(["file content"], "test.txt", { type: "text/plain" });
    const input = screen.getByTestId("file-input");

    act(() => {
      fireEvent.change(input, { target: { files: [file] } });
    });

    await waitFor(() => {
      expect(screen.getByTestId("file-0")).toBeInTheDocument();
    });

    act(() => {
      fireEvent.click(screen.getByTestId("remove-file-0"));
    });

    await waitFor(() => {
      expect(screen.queryByTestId("file-0")).not.toBeInTheDocument();
    });
  });

  test("processes file content and displays transactions", async () => {
    render(<FileUploader />);

    const file = new File(["file content"], "test.txt", { type: "text/plain" });
    const input = screen.getByTestId("file-input");

    act(() => {
      fireEvent.change(input, { target: { files: [file] } });
    });

    await waitFor(() => {
      expect(screen.getByTestId("transaction-0")).toBeInTheDocument();
      expect(screen.getByTestId("transaction-0")).toHaveTextContent("Store A");
      expect(fileToTransactions).toHaveBeenCalledWith("mock file content");
    });
  });

  test("disables drag and drop when files are present", async () => {
    render(<FileUploader />);

    const file = new File(["file content"], "test.txt", { type: "text/plain" });
    const input = screen.getByTestId("file-input");

    act(() => {
      fireEvent.change(input, { target: { files: [file] } });
    });

    await waitFor(() => {
      expect(screen.getByTestId("file-0")).toBeInTheDocument();
      expect(screen.getByText("disabled: true")).toBeInTheDocument();
    });
  });

  test("clears transactions when file is removed", async () => {
    render(<FileUploader />);

    const file = new File(["file content"], "test.txt", { type: "text/plain" });
    const input = screen.getByTestId("file-input");

    act(() => {
      fireEvent.change(input, { target: { files: [file] } });
    });

    await waitFor(() => {
      expect(screen.getByTestId("transaction-0")).toBeInTheDocument();
    });

    act(() => {
      fireEvent.click(screen.getByTestId("remove-file-0"));
    });

    await waitFor(() => {
      expect(screen.queryByTestId("transaction-0")).not.toBeInTheDocument();
    });
  });

  test("handles multiple files", async () => {
    render(<FileUploader />);

    const file1 = new File(["content 1"], "test1.txt", { type: "text/plain" });
    const input = screen.getByTestId("file-input");

    act(() => {
      fireEvent.change(input, { target: { files: [file1] } });
    });

    await waitFor(() => {
      expect(screen.getByTestId("file-0")).toBeInTheDocument();
    });

    const file2 = new File(["content 2"], "test2.txt", { type: "text/plain" });

    act(() => {
      fireEvent.change(input, { target: { files: [file2] } });
    });

    await waitFor(() => {
      expect(screen.getByTestId("file-1")).toBeInTheDocument();
      expect(screen.getByTestId("file-0")).toHaveTextContent("test1.txt");
      expect(screen.getByTestId("file-1")).toHaveTextContent("test2.txt");
    });
  });

  test("matches snapshot", () => {
    const { container } = render(<FileUploader />);
    expect(container).toMatchSnapshot();
  });
});
