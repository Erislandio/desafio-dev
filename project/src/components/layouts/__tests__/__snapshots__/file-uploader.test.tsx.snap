// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileUploader Component matches snapshot 1`] = `
<div>
  <div>
    <div
      data-testid="drag-drop-component"
    >
      <button
        data-testid="simulate-drop"
      >
        Simulate Drop
      </button>
      <button
        data-testid="simulate-drag-over"
      >
        Simulate Drag Over
      </button>
      <button
        data-testid="simulate-drag-leave"
      >
        Simulate Drag Leave
      </button>
      <input
        data-testid="file-input"
        type="file"
      />
      <div>
        isDragging: 
        false
      </div>
      <div>
        uploading: 
        false
      </div>
      <div>
        disabled: 
        false
      </div>
    </div>
    <div
      data-testid="files-list-component"
    />
    <div
      data-testid="transaction-table-component"
    />
    <div
      class="flex justify-end mt-4"
    >
      <button
        data-testid="upload-button"
        disabled=""
      >
        Enviar
      </button>
    </div>
  </div>
</div>
`;
